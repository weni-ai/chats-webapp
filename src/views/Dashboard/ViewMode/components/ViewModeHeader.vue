<template>
  <header
    class="view-mode__header"
    v-if="viewedAgent"
  >
    <h1 class="title">
      <UnnnicIcon
        icon="visibility"
        size="md"
      />
      {{ $t('dashboard.view-mode.title', { viewedAgent }) }}
    </h1>
    <div
      class="close-button"
      @click="closeViewMode"
      @keypress.enter="closeViewMode"
    >
      <UnnnicToolTip
        :text="$t('dashboard.view-mode.close')"
        enabled
        side="left"
      >
        <UnnnicIcon
          size="sm"
          icon="close-1"
        />
      </UnnnicToolTip>
    </div>
  </header>
</template>
<script>
export default {
  name: 'ViewModeHeader',

  props: {
    viewedAgent: {
      required: true,
      type: String,
    },
  },

  methods: {
    async closeViewMode() {
      await this.$store.dispatch('chats/rooms/setActiveRoom', null);
      this.$router.push({ name: 'dashboard.manager' });
    },
  },
};
</script>

<style lang="scss" scoped>
.view-mode__header {
  position: absolute;
  top: 0;
  left: 0;

  display: flex;
  justify-content: space-between;
  align-items: center;

  padding: $unnnic-spacing-stack-xs 0;

  width: 100%;

  background: $unnnic-color-brand-weni;
  color: $unnnic-color-neutral-black;

  cursor: default;

  .title {
    display: flex;
    gap: $unnnic-spacing-inline-xs;
    justify-content: center;

    font-size: $unnnic-font-size-body-lg;
    font-weight: $unnnic-font-weight-regular;

    flex: 1;
  }

  .close-button {
    margin-right: calc(
      ($unnnic-font-size-body-lg / 100 * 50) + $unnnic-spacing-stack-xs
    );
    cursor: pointer;

    .unnnic-tooltip {
      display: flex;
    }
  }
}
</style>
