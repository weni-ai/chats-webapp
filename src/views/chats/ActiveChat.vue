<template>
  <chats-layout>
    <section v-if="!!room" class="active-chat">
      <chat-header
        :room="room"
        :closeButtonTooltip="$t('chats.end')"
        @close="openModalCloseChat"
        @show-contact-info="componentInAsideSlot = 'contactInfo'"
        @open-select-flow="
          componentInAsideSlot =
            componentInAsideSlot === 'layoutTemplateMessage' ? '' : 'layoutTemplateMessage'
        "
        :alert="!this.room.is_24h_valid"
      />
      <chat-messages
        :room="room"
        :messages="messages"
        class="messages"
        @show-contact-info="componentInAsideSlot = 'contactInfo'"
        ref="chatMessages"
        @scrollTop="searchForMoreMessages"
      />

      <div v-if="isMessageEditorVisible && !room.is_waiting" class="message-editor">
        <message-editor
          ref="message-editor"
          v-model="editorMessage"
          :audio.sync="audioMessage"
          @show-quick-messages="
            componentInAsideSlot = componentInAsideSlot === 'quickMessages' ? '' : 'quickMessages'
          "
          @send-message="sendMessage"
          @send-audio="sendAudio"
          @upload="sendFileMessage($event)"
          :loadingValue="totalValue"
          :loading="isLoading"
        />
      </div>

      <div v-if="!room.user" class="get-chat-button-container">
        <unnnic-button
          class="get-chat-button"
          :text="$t('chats.get_chat')"
          type="secondary"
          @click="isGetChatConfirmationModalOpen = true"
        />
      </div>

      <section v-if="isRoomClassifierVisible" class="chat-classifier">
        <chat-classifier
          v-model="tags"
          :tags="sectorTags"
          label="Por favor, classifique o atendimento:"
        >
          <template #actions>
            <unnnic-button :text="$t('confirm')" type="secondary" size="small" @click="closeRoom" />
          </template>
        </chat-classifier>
      </section>
    </section>

    <unnnic-modal
      v-if="room"
      :showModal="isCloseChatModalOpen"
      @close="isCloseChatModalOpen = false"
      :text="$t('chats.end')"
      :description="$t('chats.end_confirmation', { name: room.contact.name })"
      modal-icon="alert-circle-1"
      scheme="feedback-yellow"
    >
      <template #options>
        <unnnic-button :text="$t('cancel')" type="terciary" @click="isCloseChatModalOpen = false" />
        <unnnic-button :text="$t('confirm')" type="secondary" @click="classifyRoom" />
      </template>
    </unnnic-modal>

    <unnnic-modal
      v-if="room"
      :showModal="isGetChatConfirmationModalOpen"
      @close="isGetChatConfirmationModalOpen = false"
      :text="$t('chats.get_chat_question')"
      :description="`Confirme se deseja realizar o atendimento de ${room.contact.name}`"
      modal-icon="messages-bubble-1"
      scheme="neutral-darkest"
    >
      <template #options>
        <unnnic-button
          :text="$t('cancel')"
          type="terciary"
          @click="isGetChatConfirmationModalOpen = false"
        />
        <unnnic-button :text="$t('confirm')" type="secondary" @click="takeRoom" />
      </template>
    </unnnic-modal>

    <template #aside>
      <component :is="sidebarComponent.name" v-on="sidebarComponent.listeners" />
    </template>
    <modal-close-chat v-if="showCloseModal" @close="closeModalCloseChat" :room="room" />
  </chats-layout>
</template>

<script>
import { mapState, mapGetters } from 'vuex';

import ChatsLayout from '@/layouts/ChatsLayout';

import ChatHeader from '@/components/chats/chat/ChatHeader';
import ChatMessages from '@/components/chats/chat/ChatMessages';
import ContactInfo from '@/components/chats/ContactInfo';
import MessageEditor from '@/components/chats/MessageEditor';
import ChatClassifier from '@/components/chats/ChatClassifier';
import QuickMessages from '@/components/chats/QuickMessages';
import ModalCloseChat from '@/views/chats/ModalCloseChat.vue';
import LayoutTemplateMessage from '@/components/chats/TemplateMessages/LayoutTemplateMessage';

import Room from '@/services/api/resources/chats/room';
import Queue from '@/services/api/resources/settings/queue';

export default {
  name: 'ActiveChat',

  components: {
    ChatHeader,
    ChatMessages,
    ContactInfo,
    QuickMessages,
    ChatsLayout,
    MessageEditor,
    ChatClassifier,
    ModalCloseChat,
    LayoutTemplateMessage,
  },

  props: {
    id: {
      type: String,
      default: '',
    },
  },

  data: () => ({
    /**
     * @type {HTMLAudioElement}
     */
    audioMessage: null,
    componentInAsideSlot: '',
    editorMessage: '',
    isCloseChatModalOpen: false,
    tags: [],
    sectorTags: [],
    isGetChatConfirmationModalOpen: false,
    isRoomClassifierVisible: false,
    totalValue: undefined,
    isLoading: false,
    page: 0,
    limit: 20,
    showCloseModal: false,
    showAlertForLastMessage: false,
  }),

  computed: {
    ...mapState({
      room: (state) => state.rooms.activeRoom,
      me: (state) => state.profile.me,
      hasNext: (state) => state.rooms.hasNext,
      listRoomHasNext: (state) => state.rooms.listRoomHasNext,
    }),
    ...mapGetters('rooms', {
      messages: 'groupedActiveRoomsMessage',
    }),
    isMessageEditorVisible() {
      return (
        !this.isRoomClassifierVisible &&
        this.room.is_active &&
        this.room.is_24h_valid &&
        !this.room.wating_answer &&
        !!this.room.user
      );
    },
    sidebarComponent() {
      return this.sidebarComponents[this.componentInAsideSlot] || {};
    },
    sidebarComponents() {
      return {
        quickMessages: {
          name: QuickMessages.name,
          listeners: {
            close: () => {
              this.componentInAsideSlot = '';
            },
            'select-quick-message': (quickMessage) => {
              this.editorMessage = quickMessage.text;
            },
          },
        },
        contactInfo: {
          name: ContactInfo.name,
          listeners: {
            close: () => {
              this.componentInAsideSlot = '';
            },
          },
        },
        layoutTemplateMessage: {
          name: LayoutTemplateMessage.name,
          listeners: {
            close: () => {
              this.componentInAsideSlot = '';
            },
            contact: this.room,
          },
        },
      };
    },
  },

  methods: {
    async classifyRoom() {
      this.isRoomClassifierVisible = true;
      this.isCloseChatModalOpen = false;
      const response = await Queue.tags(this.room.queue.uuid);
      this.sectorTags = response.results;
    },
    async takeRoom() {
      await Room.take(this.room.uuid, this.me.email);
      this.isGetChatConfirmationModalOpen = false;
      this.setActiveRoom(this.room.uuid);
    },
    async closeRoom() {
      // if (this.tags.length === 0) return;
      const { uuid } = this.room;

      const tags = this.tags.map((tag) => tag.uuid);
      await Room.close(uuid, tags);
      this.$router.replace({ name: 'home' });
      this.$store.dispatch('rooms/removeRoom', uuid);
    },
    scrollMessagesToBottom() {
      if (!this.$refs.chatMessages) return;
      this.$refs.chatMessages.$el.scrollTop = this.$refs.chatMessages.$el.scrollHeight;
    },
    async setActiveRoom(uuid) {
      const room = this.$store.getters['rooms/getRoomById'](uuid);
      if (!room) this.$router.push({ name: 'home' });
      await this.$store.dispatch('rooms/setActiveRoom', room);
      this.componentInAsideSlot = '';
      this.page = 0;
    },
    async getRoomMessages(concat) {
      this.isLoading = true;

      try {
        await this.$store.dispatch('rooms/getActiveRoomMessages', {
          offset: this.page * this.limit,
          concat,
          limit: this.limit,
        });
        // this.$nextTick(this.scrollMessagesToBottom);
        this.isRoomClassifierVisible = false;
        this.isLoading = false;
        this.dateOfLastMessage();
      } catch (error) {
        this.isLoading = false;
        console.log(error);
      }
    },

    searchForMoreMessages() {
      if (this.hasNext) {
        this.page += 1;
        this.getRoomMessages(true);
      }
    },
    async sendFileMessage(files) {
      try {
        const loadingFiles = {};
        const updateLoadingFiles = (messageUuid, progress) => {
          loadingFiles[messageUuid] = progress;
          this.totalValue =
            Object.values(loadingFiles).reduce((acc, value) => acc + value) /
            Object.keys(loadingFiles).length;
        };
        await this.$store.dispatch('rooms/sendMedias', { files, updateLoadingFiles });
        this.totalValue = undefined;
        this.scrollMessagesToBottom();
      } catch (e) {
        console.error('O upload de alguns arquivos pode não ter sido concluído');
      }
    },

    async sendMessage() {
      const message = this.editorMessage.trim();
      if (!message) return;

      await this.$store.dispatch('rooms/sendMessage', message);

      this.scrollMessagesToBottom();
      this.editorMessage = '';
    },
    async sendAudio() {
      if (!this.audioMessage) return;
      const loadingFiles = {};
      const updateLoadingFiles = (messageUuid, progress) => {
        loadingFiles[messageUuid] = progress;
        this.totalValue =
          Object.values(loadingFiles).reduce((acc, value) => acc + value) /
          Object.keys(loadingFiles).length;
      };
      const response = await fetch(this.audioMessage.src);
      const blob = await response.blob();
      const audio = new File([blob], `${Date.now().toString()}.mp3`, { type: 'audio/mpeg3' });
      await this.$store.dispatch('rooms/sendMedias', { files: [audio], updateLoadingFiles });
      this.totalValue = undefined;
      this.scrollMessagesToBottom();
      this.$refs['message-editor'].clearAudio();
      this.audioMessage = null;
    },
    getTodayDate() {
      return new Intl.DateTimeFormat('pt-BR', {
        dateStyle: 'short',
      }).format(new Date());
    },
    openModalCloseChat() {
      this.showCloseModal = true;
    },

    closeModalCloseChat() {
      this.showCloseModal = false;
    },

    dateOfLastMessage() {
      if (!this.room) return;
      if (!this.room.is_24h_valid) {
        this.showAlertForLastMessage = true;
      } else {
        this.showAlertForLastMessage = false;
      }
    },
  },

  watch: {
    room(newValue) {
      if (!newValue) this.componentInAsideSlot = '';
      if (newValue) this.editorMessage = '';
    },
    id: {
      immediate: true,
      async handler() {
        if (this.$store.state.rooms.newMessagesByRoom[this.id]) {
          this.$delete(this.$store.state.rooms.newMessagesByRoom, this.id);
        }

        await this.setActiveRoom(this.id);
        this.getRoomMessages();
      },
    },

    messages() {
      this.$nextTick(this.scrollMessagesToBottom);
    },
  },
};
</script>

<style lang="scss" scoped>
.active-chat {
  display: flex;
  flex-direction: column;
  height: 100%;
  padding: {
    top: $unnnic-spacing-inset-md;
    bottom: $unnnic-spacing-inset-sm;
    left: $unnnic-spacing-inset-md;
  }

  .messages {
    overflow-y: auto;
    padding-right: $unnnic-spacing-inset-sm;
    margin: $unnnic-spacing-inline-sm 0 $unnnic-spacing-inline-sm;
  }

  .chat-classifier {
    margin-top: auto;
    margin-left: -$unnnic-spacing-inline-md;
    margin-bottom: -$unnnic-spacing-inline-sm;
  }

  .message-editor {
    margin-right: $unnnic-spacing-inline-sm;
    margin-top: auto;
  }

  .get-chat-button-container {
    margin-top: auto;
    margin-right: $unnnic-spacing-inline-sm;

    .get-chat-button {
      width: 100%;
    }
  }
}
</style>
