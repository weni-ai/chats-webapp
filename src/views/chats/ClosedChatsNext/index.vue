<template>
  <closed-chats-header-loading v-if="isLoadingHeader" />
  <div class="closed-chats" v-else>
    <header v-if="project">
      <unnnic-chats-header
        :title="project.name"
        :subtitle="$t('chats.closed_chats.project_history')"
        avatarIcon="task-list-clock-1"
        :crumbs="crumbs"
        :close="backToHome"
      />
      <room-header-loading v-show="roomId && isLoadingSelectedRoom" />
      <unnnic-chats-header
        v-if="!isLoadingSelectedRoom && selectedRoom"
        :title="selectedRoom.contact.name"
        :avatarName="selectedRoom.contact.name"
        :close="() => $router.push({ name: 'closed-rooms' })"
      />
    </header>
    <main>
      <section v-if="roomId" class="closed-chats__selected-chat">
        <room-loading v-show="isLoadingSelectedRoom" only-messages />
        <div v-show="!isLoadingSelectedRoom && selectedRoom">chatMessages</div>
        <contact-info is-history :contact="selectedRoom" @close="() => {}" />
      </section>

      <closed-chats-rooms-table v-else :project="project" />
    </main>
  </div>
</template>

<script>
import ProjectApi from '@/services/api/resources/settings/project';
import History from '@/services/api/resources/chats/history';
import Message from '@/services/api/resources/chats/message';

import ContactInfo from '@/components/chats/ContactInfo';
import ClosedChatsHeaderLoading from '@/views/loadings/ClosedChats/ClosedChatsHeader.vue';
import RoomHeaderLoading from '@/views/loadings/RoomHeader.vue';
import RoomLoading from '@/views/loadings/Room.vue';
import ClosedChatsRoomsTable from './RoomsTable';

export default {
  name: 'ClosedChatsNext',

  components: {
    ClosedChatsHeaderLoading,
    RoomHeaderLoading,
    RoomLoading,
    ContactInfo,
    ClosedChatsRoomsTable,
    // TagGroup,
  },

  props: {
    roomId: {
      type: String,
      default: '',
    },
  },

  data: () => ({
    isLoadingHeader: true,
    isLoadingSelectedRoom: false,

    crumbs: [
      {
        name: 'Chats',
        path: 'home',
      },
    ],
    project: null,
  }),

  created() {
    this.projectInfo();
    this.crumbs.push({
      name: this.$t('chats.closed_chats.history'),
      path: 'closed-rooms',
    });
  },

  methods: {
    backToHome() {
      this.$router.push({ name: 'home' });
    },

    async projectInfo() {
      const project = await ProjectApi.getInfo();
      this.project = project.data;
      this.isLoadingHeader = false;
    },
  },

  watch: {
    roomId: {
      immediate: true,
      async handler(roomId) {
        if (roomId) {
          this.isLoadingSelectedRoom = true;

          this.crumbs.push({
            name: 'Nome do usu√°rio',
            path: 'closed-rooms/:roomId',
          });
          const responseRoom = await History.getHistoryContactRoom({ room: roomId });
          const responseRoomMessages = await Message.getByContact(roomId, 0, 20);
          console.log(responseRoomMessages);
          this.selectedRoom = responseRoom;
          // this.selectedRoomMessages = responseRoomMessages;

          this.isLoadingSelectedRoom = false;
        }
      },
    },
  },
};
</script>

<style lang="scss" scoped>
.closed-chats {
  display: flex;
  flex-direction: column;

  height: 100vh;
  width: 100vw;

  overflow: hidden;

  main {
    height: 100%;
    overflow: hidden;
  }

  &__selected-chat {
    display: grid;
    grid-template-columns: 9fr 3fr;

    :deep(.unnnic-chats-header) {
      display: none;
    }
  }
}
</style>
