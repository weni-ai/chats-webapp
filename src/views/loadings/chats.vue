<template>
  <div class="side-bar">
    <div class="unnnic-grid-giant" style="padding: 12px 24px">
      <div class="unnnic-grid-span-10" style="display: flex; justify-content: space-between">
        <div style="margin-top: 25px">
          <unnnic-skeleton-loading height="20px" width="90px" />
        </div>
        <div>
          <unnnic-skeleton-loading height="45px" width="670px" />
        </div>
      </div>
    </div>
    <div class="unnnic-grid-giant" style="padding: 12px 24px">
      <div class="unnnic-grid-span-12" style="display: flex; justify-content: space-between">
        <div>
          <unnnic-skeleton-loading height="45px" width="350px" />
        </div>
        <div>
          <unnnic-skeleton-loading height="45px" width="200px" />
        </div>
        <div>
          <unnnic-skeleton-loading height="45px" width="200px" />
        </div>
        <div>
          <unnnic-skeleton-loading height="45px" width="200px" />
        </div>
        <div>
          <unnnic-skeleton-loading height="45px" width="200px" />
        </div>
      </div>
    </div>
    <div class="unnnic-grid-giant">
      <div class="unnnic-grid-span-1" style="display: flex; flex-direction: column">
        <div class="skeleton-line"><unnnic-skeleton-loading height="45px" width="350px" /></div>
        <div class="skeleton-line"><unnnic-skeleton-loading height="45px" width="350px" /></div>
        <div class="skeleton-line"><unnnic-skeleton-loading height="20px" width="90px" /></div>
        <div class="skeleton-line"><unnnic-skeleton-loading height="45px" width="350px" /></div>
        <div class="skeleton-line"><unnnic-skeleton-loading height="45px" width="350px" /></div>
        <div class="skeleton-line"><unnnic-skeleton-loading height="45px" width="350px" /></div>
        <div class="skeleton-line"><unnnic-skeleton-loading height="45px" width="350px" /></div>
        <div class="skeleton-line"><unnnic-skeleton-loading height="45px" width="350px" /></div>
        <div class="skeleton-line"><unnnic-skeleton-loading height="45px" width="350px" /></div>
        <div class="skeleton-line"><unnnic-skeleton-loading height="45px" width="350px" /></div>
        <div class="skeleton-line"><unnnic-skeleton-loading height="45px" width="350px" /></div>
      </div>
    </div>
    <div class="page"></div>
  </div>
</template>

<script>
export default {
  name: 'ProjectsLoading',
};
</script>

<style scoped lang="scss">
.side-bar {
  position: absolute;
  left: 0%;
  right: 0%;
  top: 0%;
  bottom: 0.14%;
  background-color: #ffffff;
  opacity: 2;
  border-radius: 4px;
}
.page {
  background: linear-gradient(90deg, #e8f4f4 0%, #ffffff 100%);
  position: absolute;
  width: 50%;
  height: 80%;
  left: 400px;
  top: 165px;
}

.skeleton-line {
  padding: 7px 0px;
}
</style>
