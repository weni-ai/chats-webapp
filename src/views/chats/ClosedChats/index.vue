<template>
  <div class="closed-chats">
    <ClosedChatsHeaderLoading v-if="isLoadingHeader" />
    <header v-if="!isLoadingHeader && project">
      <UnnnicChatsHeader
        :title="project.name"
        :subtitle="$t('chats.closed_chats.project_history')"
        avatarIcon="history"
        :crumbs="crumbs"
        :close="backToHome"
        @crumbClick="handlerCrumbClick"
        :size="closedChatsHeaderSize"
      />
      <ChatHeaderLoading v-show="roomId && isLoadingSelectedRoom" />
      <UnnnicChatsHeader
        v-show="!isLoadingSelectedRoom"
        v-if="selectedRoom"
        :title="selectedRoom.contact.name"
        :avatarName="selectedRoom.contact.name"
      />
    </header>
    <main>
      <section
        v-if="roomId"
        class="closed-chats__selected-chat"
      >
        <RoomMessages />
        <ContactInfo
          isHistory
          :closedRoom="selectedRoom"
          @close="() => {}"
        />
      </section>

      <ClosedChatsRoomsTable v-else />
    </main>
  </div>
</template>

<script>
import isMobile from 'is-mobile';

import { mapActions, mapState } from 'pinia';
import { useRooms } from '@/store/modules/chats/rooms';
import { useRoomMessages } from '@/store/modules/chats/roomMessages';
import { useConfig } from '@/store/modules/config';
import History from '@/services/api/resources/chats/history';

import RoomMessages from '@/components/chats/chat/RoomMessages.vue';
import ContactInfo from '@/components/chats/ContactInfo/index.vue';
import ClosedChatsHeaderLoading from '@/views/loadings/ClosedChats/ClosedChatsHeader.vue';
import ChatHeaderLoading from '@/views/loadings/chat/ChatHeader.vue';
import ClosedChatsRoomsTable from './RoomsTable.vue';

export default {
  name: 'ClosedChats',

  components: {
    ClosedChatsHeaderLoading,
    ChatHeaderLoading,
    ContactInfo,
    ClosedChatsRoomsTable,
    RoomMessages,
  },

  props: {
    roomId: {
      type: String,
      default: '',
    },
  },

  data: () => ({
    isMobile: isMobile(),

    isLoadingHeader: true,
    isLoadingSelectedRoom: false,

    crumbs: [
      {
        name: 'Chats',
        path: 'home',
      },
    ],

    selectedRoom: null,
    selectedRoomsUuids: null,
  }),

  async created() {
    this.crumbs.push({
      name: this.$t('chats.closed_chats.history'),
      path: 'closed-rooms',
    });
  },

  computed: {
    ...mapState(useConfig, ['project']),
    ...mapState(useRoomMessages, ['roomMessagesNext']),

    closedChatsHeaderSize() {
      return this.isMobile ? 'small' : 'large';
    },
  },

  methods: {
    ...mapActions(useRooms, ['setActiveRoom']),
    ...mapActions(useRoomMessages, ['getRoomMessages', 'resetRoomMessages']),

    backToHome() {
      this.$router.push({ name: 'home' });
    },

    handlerCrumbClick(crumb) {
      if (crumb.name === this.selectedRoom?.contact.name) return;

      this.selectedRoom = null;

      const index = this.crumbs.findIndex((item) => item.path === crumb.path);
      this.crumbs = this.crumbs.slice(0, index + 1);

      this.$router.push({ name: crumb.path });
    },

    async chatScrollTop() {
      if (this.roomMessagesNext) {
        this.getHistoryContactRoomMessages();
      } else {
        // const roomUuidIndex = this.selectedRoomsUuids?.findIndex(
        //   (room) => room.uuid === this.roomId,
        // );
        // const previousRoom = this.selectedRoomsUuids[roomUuidIndex - 1];
        //
        // if (previousRoom) {
        //   const responseRoom = await History.getHistoryContactRoom({ room: previousRoom.uuid });
        //   await this.setActiveRoom(responseRoom);
        //   this.getHistoryContactRoomMessages();
        //   this.selectedRoom = responseRoom;
        // }
      }
    },

    async getHistoryContactRoomMessages() {
      await this.getRoomMessages();
    },
  },

  watch: {
    roomId: {
      immediate: true,
      async handler(roomId) {
        if (!roomId) {
          await this.setActiveRoom(null);
          await this.resetRoomMessages();
        }
        if (roomId) {
          this.isLoadingSelectedRoom = true;

          const responseRoom = await History.getHistoryContactRoom({
            room: roomId,
          });

          const STATUS_NOT_FOUND = 404;
          if (responseRoom.status === STATUS_NOT_FOUND) {
            this.$router.push({ name: 'closed-rooms' });
            return;
          }

          this.crumbs.push({
            name: responseRoom.contact.name,
            path: 'closed-rooms/:roomId',
          });

          this.selectedRoom = responseRoom;
          await this.setActiveRoom(this.selectedRoom);
          await this.getHistoryContactRoomMessages();
          const responseRoomUuids = await History.getHistoryContactRoomsUuids({
            external_id: responseRoom.contact.external_id,
          });
          this.selectedRoomsUuids = responseRoomUuids.results;

          this.isLoadingSelectedRoom = false;
        }
      },
    },
    project: {
      immediate: true,
      handler(newProject) {
        if (newProject) {
          this.isLoadingHeader = false;
        }
      },
    },
  },
};
</script>

<style lang="scss" scoped>
.closed-chats {
  display: flex;
  flex-direction: column;

  height: 100vh;
  width: 100vw;

  overflow: hidden;

  main {
    height: 100%;
    overflow: hidden;
  }

  &__selected-chat {
    display: grid;
    grid-template-columns: 9fr 3fr;
    grid-template-rows: 100%;

    padding-left: $unnnic-spacing-sm;

    height: 100%;

    background-color: $unnnic-color-background-carpet;

    :deep(.unnnic-chats-header) {
      display: none;
    }
  }
}
</style>
