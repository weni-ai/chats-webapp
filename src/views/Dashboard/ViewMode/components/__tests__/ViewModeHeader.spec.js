import { afterAll, beforeAll, describe, expect, it, vi } from 'vitest';
import { mount, config } from '@vue/test-utils';
import { createPinia, setActivePinia } from 'pinia';
import ViewModeHeader from '@/views/Dashboard/ViewMode/components/ViewModeHeader.vue';
import { useRooms } from '@/store/modules/chats/rooms';
import i18n from '@/plugins/i18n';

beforeAll(() => {
  config.global.plugins = config.global.plugins.filter(
    (plugin) => plugin !== i18n,
  );
});

afterAll(() => {
  if (!config.global.plugins.includes(i18n)) {
    config.global.plugins.push(i18n);
  }
});

describe('ViewModeHeader', () => {
  const createWrapper = (props = {}, routeParams = {}) => {
    const pinia = createPinia();
    setActivePinia(pinia);

    const mockRouter = { push: vi.fn() };
    const mockRoute = { params: routeParams };

    return mount(ViewModeHeader, {
      props: { viewedAgent: 'John Doe', ...props },
      global: {
        plugins: [pinia],
        mocks: {
          $t: (key, params) => `${key}_${params?.viewedAgent || ''}`,
          $router: mockRouter,
          $route: mockRoute,
        },
        stubs: {
          UnnnicIcon: {
            template: '<span data-testid="icon" :data-icon="icon"></span>',
            props: ['icon', 'size'],
          },
          UnnnicToolTip: {
            template: '<div data-testid="tooltip"><slot /></div>',
            props: ['text', 'enabled', 'side'],
          },
        },
      },
    });
  };

  describe('Props and Initial State', () => {
    it('should handle props and initialize correctly', () => {
      const wrapper = createWrapper({ viewedAgent: 'Test Agent' });
      expect(wrapper.props('viewedAgent')).toBe('Test Agent');
      expect(wrapper.vm.isInsight).toBe(false);
    });
  });

  describe('Conditional Rendering', () => {
    it('should render header when viewedAgent is provided', () => {
      const wrapper = createWrapper({ viewedAgent: 'John Doe' });

      expect(wrapper.find('[data-testid="view-mode-header"]').exists()).toBe(
        true,
      );
      expect(wrapper.find('[data-testid="header-title"]').exists()).toBe(true);
      expect(wrapper.find('[data-testid="visibility-icon"]').exists()).toBe(
        true,
      );
      expect(wrapper.find('[data-testid="close-button"]').exists()).toBe(true);
      expect(wrapper.find('[data-testid="close-tooltip"]').exists()).toBe(true);
      expect(wrapper.find('[data-testid="close-icon"]').exists()).toBe(true);
    });

    it('should not render header when viewedAgent is falsy', () => {
      const wrapperNull = createWrapper({ viewedAgent: null });
      const wrapperEmpty = createWrapper({ viewedAgent: '' });
      const wrapperUndefined = createWrapper({ viewedAgent: undefined });

      expect(
        wrapperNull.find('[data-testid="view-mode-header"]').exists(),
      ).toBe(false);
      expect(
        wrapperEmpty.find('[data-testid="view-mode-header"]').exists(),
      ).toBe(false);
      expect(
        wrapperUndefined.find('[data-testid="view-mode-header"]').exists(),
      ).toBe(false);
    });

    it('should display correct internationalized title', () => {
      const wrapper = createWrapper({ viewedAgent: 'Jane Smith' });
      const title = wrapper.find('[data-testid="header-title"]');

      expect(title.text()).toContain('dashboard.view-mode.title_Jane Smith');
    });
  });

  describe('Functionality', () => {
    it('should handle closeViewMode method with regular navigation', async () => {
      const wrapper = createWrapper();
      const setActiveRoomSpy = vi.spyOn(wrapper.vm, 'setActiveRoom');

      await wrapper.vm.closeViewMode();

      expect(setActiveRoomSpy).toHaveBeenCalledWith(null);
      expect(wrapper.vm.$router.push).toHaveBeenCalledWith({
        name: 'dashboard.manager',
      });
    });

    it('should handle closeViewMode with insights module postMessage', async () => {
      const postMessageSpy = vi.spyOn(window.parent, 'postMessage');
      const wrapper = createWrapper({}, { oldModule: 'insights' });

      await wrapper.vm.closeViewMode();

      expect(postMessageSpy).toHaveBeenCalledWith(
        { event: 'redirect', path: 'insights:init' },
        '*',
      );
    });

    it('should handle click and keypress events on close button', async () => {
      const wrapper = createWrapper();
      const closeViewModeSpy = vi.spyOn(wrapper.vm, 'closeViewMode');
      const closeButton = wrapper.find('[data-testid="close-button"]');

      await closeButton.trigger('click');
      expect(closeViewModeSpy).toHaveBeenCalledTimes(1);

      await closeButton.trigger('keypress.enter');
      expect(closeViewModeSpy).toHaveBeenCalledTimes(2);
    });

    it('should integrate with rooms store correctly', () => {
      const wrapper = createWrapper();
      expect(typeof wrapper.vm.setActiveRoom).toBe('function');

      const roomsStore = useRooms();
      expect(roomsStore).toBeDefined();
    });
  });

  describe('Snapshot', () => {
    it('should match snapshot', () => {
      const wrapper = createWrapper({ viewedAgent: 'Test User' });
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
