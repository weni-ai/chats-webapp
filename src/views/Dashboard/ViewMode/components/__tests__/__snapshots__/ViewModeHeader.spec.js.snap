// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ViewModeHeader > Snapshot > should match snapshot 1`] = `
<header
  class="view-mode__header"
  data-testid="view-mode-header"
  data-v-d329e103=""
>
  <h1
    class="title"
    data-testid="header-title"
    data-v-d329e103=""
  >
    <span
      data-icon="visibility"
      data-testid="visibility-icon"
      data-v-d329e103=""
    />
     dashboard.view-mode.title_Test User
  </h1>
  <div
    class="close-button"
    data-testid="close-button"
    data-v-d329e103=""
  >
    <div
      class="unnnic-tooltip"
      data-testid="close-tooltip"
      data-v-8c61d974=""
      data-v-d329e103=""
    >
      
      <span
        data-icon="close-1"
        data-testid="close-icon"
        data-v-d329e103=""
      />
      
      <span
        class="unnnic-tooltip-label unnnic-tooltip-label-left"
        data-testid="tooltip-label"
        data-v-8c61d974=""
      >
        
        
        dashboard.view-mode.close_ 
        <br
          data-v-8c61d974=""
        />
        
        
        <!---->
      </span>
    </div>
  </div>
</header>
`;
